daemon, node: Remove old, discarded router IPs from `cilium_host`

christarazi opened the pull request:
In the previous commit (referenced below), we forgot to remove the old
router IPs from the actual interface (`cilium_host`). This caused
connectivity issues in user environments where the discarded, stale IPs
were reassigned to pods, causing the ipcache entries for those IPs to
have `remote-node` identity.

To fix this, we remove all IPs from the `cilium_host` interface that
weren't restored during the router IP restoration process. This step
correctly finalizes the restoration process for router IPs.

Fixes: ff63b0775c0 ("daemon, node: Fix faulty router IP restoration
logic")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon, node: Remove old, discarded router IPs from `cilium_host`

In the previous commit (referenced below), we forgot to remove the old
router IPs from the actual interface (`cilium_host`). This caused
connectivity issues in user environments where the discarded, stale IPs
were reassigned to pods, causing the ipcache entries for those IPs to
have `remote-node` identity.

To fix this, we remove all IPs from the `cilium_host` interface that
weren't restored during the router IP restoration process. This step
correctly finalizes the restoration process for router IPs.

Fixes: ff63b0775c0 ("daemon, node: Fix faulty router IP restoration
logic")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -276,7 +277,48 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+	if err := removeOldRouterState(
+		node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr),
+	); err != nil {
+		log.WithError(err).Warn("Failed to remove old router IPs from cilium_host. Manual intervention is required to remove the old IPs.")
```
How does the user know which IPs are good or which ones are stale if they see this log? Do we know that the IP(s) will show up in `err`?

christarazi replied:
I've added the restored IP to the log msg and clarified that all other IPs should be removed.

joestringer replied:
I think that the latest version adds the intended IP to restore into the message, but I was also wondering about the IPs that were not removed correctly. I just didn't dig deep enough into `err` to understand whether `err` would be something like `Permission Denied` or `Failed to remove IP %s from device %s: Permission Denied`.

christarazi replied:
Ah I see. Should be resolved now as I format the error to include the IPs.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -276,7 +277,48 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+	}
+}
+
+func removeOldRouterState(restoredIP net.IP) error {
```
In a dual-stack environment, I'd expect there to be two restored IPs: one for IPv4 and one for IPv6. The caller is passing in the result of `RestoreHostIPs()`, which implies multiple yet there is only one here. Is there some other constraint somewhere that is ensuring there will only be exactly one IP?

joestringer replied:
Ah, hmm, walking up to the callers of `restoreCiliumHostIPs()` up at the daemon level, this path will be called twice in the dual-stack case. Given that this code will only list addresses from one family at a time, this should not cause problems.

joestringer commented on pkg/node/address.go:
```diff
@@ -408,6 +410,8 @@ func RestoreHostIPs(ipv6 bool, fromK8s, fromFS net.IP, cidr *cidr.CIDR) {
 		setter(ip)
 	}
+
+	return ip
```
Structurally it seems almost like the `errDoesNotBelong` case above is the case we're interested in here for the IP to remove. At the same time, I would expect the code at some point to do a walk of the IPs on the interface like the new code elsewhere in this PR does, and delete IPs that are not the "expected" one. :thinking: 

Looking back over this code, I don't really understand the switch case for `errDoesNotBelong` as rather than picking the IP and assigning that using `setter`, it will clear the address. Seems like this means that even if we have an invalid IP from the filesystem (invalid = out of CIDR range) + a valid IP from k8s, we will throw the IPs away and refuse to configure this particular family (v4/v6). This is based on following the child logic in `chooseHostIPsToRestore()`.

I'm not sure that the above is necessarily a(n actionable) criticism of this PR in particular, but this code seems like it's more nuanced than it needs to be, so I'm having trouble convincing myself it will work as expected in all cases.

christarazi replied:
I assume the case you're talking about is

```
	switch {
	case fromK8s != nil && fromFS != nil:
		if fromK8s.Equal(fromFS) {
			ip = fromK8s
		} else {
			ip = fromFS
			err = errMismatch
		}
```

If the K8s and FS IP are the same, then yes we will throw away the IP, by setting it to `nil` which will trigger code later on in the daemon to allocate a new IP. If they are NOT the same, then we will use the FS IP because we prefer local state over the K8s source. In the case that the FS IP is actually invalid (out of CIDR range), then yes we don't try to use the K8s IP. That is the missing case.

However, you mention that it shouldn't be this nuanced, but I'm not clear why it shouldn't. Shouldn't we try our best to restore the IP and ensure that whichever source it's from, that it is a valid IP given the node setup?

joestringer replied:
> However, you mention that it shouldn't be this nuanced, but I'm not clear why it shouldn't. Shouldn't we try our best to restore the IP and ensure that whichever source it's from, that it is a valid IP given the node setup?

Yes, we should. In principle I think that this is the right goal and I think that the code achieves it today for the most part.

>  In the case that the FS IP is actually invalid (out of CIDR range), then yes we don't try to use the K8s IP. That is the missing case.

I think this is my main concern. I found it a bit tricky to understand the implications of this case. Above I came to the conclusion that this shouldn't be a big deal because it would only occur if you changed the CIDR, but the fact that this case exists makes me feel like the code was written with the "good cases" in mind and the "bad cases" (eg changing CIDR) were ignored. Ideally the path would encompass all possible ways this code could be called.

More concretely, in that case, we return `nil`, then we go into `removeOldRouterState()` and iff there is >1 address assigned to the cilium_host interface, _then_ we will go and remove all IPs from the cilium_host device and leave the overall system in a weird state. The number of conditions that need to occur in order to trigger this bad state is large and I can imagine anyone having a lot of trouble trying to debug how things got to that point.

The other background thought I had is just that this code is trying to be ambivalent about which address is being used - the caller all the way up in the daemon passes exactly one address, IPv4 or IPv6. So in theory this code can just focus on the addresses that are provided to it. And yet, in each of the `removeOldRouterState()` and `RestoreHostIPs()` the code is still trying to handle both the IPv4 and IPv6 cases. (or even implicitly avoiding touching IPs from the other family). So we don't gain the benefits of abstracting away the IP family (ie that the reader doesn't have to worry about the other case), but instead the reader must understand multiple layers of functions all through this area in order to understand the way that IPv4 addresses or IPv6 addresses are handled during restore. This is for instance why I was surprised by the actual result of [this discussion thread](https://github.com/cilium/cilium/pull/17762#discussion_r741478517) where I came to the conclusion that the code is correct, but implicitly because of the interaction between the way that IP addresses are listed from the kernel and the way that this function is called with one particular address family at a time.

Similarly, from this thread when I raise that 'it seems like the `errDoesNotBelong` case seems like the case that `removeOldRouterState()` is trying to handle, I worry that we're duplicating the reaction logic across multiple different functions in ways that are subtly different. If the conditional statements don't 100% line up then it's not obvious whether this is intended because the code was thought through deeply, or because these two pieces of logic were supposed to happen in exactly the same conditions (which implies that they should be run from the same function in the same place).

I guess, as a summary the abstractions here seem a bit leaky. It's probably not going to cause problems for 90% of cases but I wish it were more "obviously correct" rather than needing the reviewer/reader to study the various interactions between 4 or 5 functions to understand whether this whole restore logic works together.

joestringer replied:
We discussed these concerns out-of-band and agreed that the potential bug raised above can be addressed in a new commit (included in the latest version). With everything in place we can consider refactoring to improve the readability of this code, but it's not as crucial and not really appropriate for this bugfix PR.

christarazi replied:
Synced offline about resolving the missing logic, which should be addressed in the latest push. I created https://github.com/cilium/cilium/issues/17771 as a followup.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -276,7 +277,48 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+
+func removeOldRouterState(restoredIP net.IP) error {
+	if restoredIP == nil {
+		return nil
```
If the calling logic decided that no IP should be assigned to the interface, is the right answer to keep all IPs on the interface? This is probably 'safer' in that it won't remove IPs that are being actively used, but on the other hand it seems like it could potentially cause the bug which this PR aims to fix, but under different conditions :thinking: .

christarazi replied:
Good point, I've addressed this in the latest push

joestringer reviewed (commented):
I think this will do the right thing in the common circumstances, although I would like to know that the log messages will include all possibly-relevant information in them (I found this a little hard to tell at face value).

Most of the discussion below is just looking over this code with fresh eyes and I found it a bit confusing for the case where the overall IP allocation CIDR range changes. Maybe we're not so worried about that case, but it looks like it could cause some hiccups if anyone did change the CIDR.

christarazi committed:
node: Add missing fallback to router IP from CiliumNode for restoration

Previously in the case that both router IPs from the filesystem and the
CiliumNode resource were available, we missed a fallback to the
CiliumNode IP, if the IP from the FS was outside the provided CIDR
range. In other words, we returned early that the FS IP does not belong
to the CIDR, without checking if the IP from the CiliumNode was a valid
fallback.

This commit adds the missing case logic and also adds more documentation
to the function.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -276,7 +277,55 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+
+		deleteIP:
+			if err := netlink.AddrDel(l, &a); err != nil {
+				errs = append(errs, err)
```
This is what I was trying to get at in the [earlier thread](https://github.com/cilium/cilium/pull/17762#discussion_r741476983):

```suggestion
				errs = append(errs, fmt.Errorf("cannot remove IP %s from device: %w", a, err))
```

joestringer reviewed (approved):
LGTM, thanks.

borkmann commented on daemon/cmd/daemon.go:
```diff
@@ -276,7 +277,55 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+	if err != nil {
+		return err
+	}
+	if len(addrs) > 1 {
```
If `len(addrs) == 1` and `restoredIP` is nil, is the assumption correct that we keep the old router ip?

borkmann replied:
If it is indeed the case, maybe a comment should be added to give some more context on why it is a valid case.

christarazi replied:
Good point @borkmann. The [diff](https://github.com/cilium/cilium/compare/a9c54b38e38507339e8dca284b248080e7a92858..587f4c19a16b22c3e2a6018aba2f9b5b6ea9e3ad) of the change is here. It is not a correct to keep the old router IP because it is not valid. We therefore should remove it from the interface.

borkmann commented on daemon/cmd/daemon.go:
```diff
@@ -274,7 +275,61 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
+			}
+			if restoredIP.Equal(a.IP) {
+				continue
+			}
```
I'm probably missing something, but why we cannot simplify the logic like the below?

pseudo-code wrt what I mean:
```
	addrs, err := netlink.AddrList(l, family)
	if err != nil {
		return err
	}

	for _, a := range addrs {
		if  restoredIP != nil && restoredIP.Equal(a.IP) {
			continue
		}
		if err := netlink.AddrDel(l, &a); err != nil {
				errs = append(errs, fmt.Errorf("failed to remove IP %s: %w", a.IP, err))
		}
	}
```

christarazi replied:
Thanks Daniel, good suggestion. I didn't notice it can be simplified. [Diff](https://github.com/cilium/cilium/compare/6844f420b23b0c51c4b67bbc791f9bfeedb38dc2..cd25bc1153a746903ff2e5f50f3c466cfd6216aa) pushed.

borkmann reviewed (approved):
thx, removeOldRouterState looks reasonable to me
